### PR TITLE
neothesia: init at unstable-2024-07-21

### PR DESCRIPTION
**Pull request by auroraanna**
## Description of changes

Added neothesia package.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by auroraanna** [comment]
Blocked by <https://github.com/PolyMeilex/Neothesia/issues/77>.

File: pkgs/applications/audio/neothesia/default.nix [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+, ffmpeg-full
+, vulkan-loader
+, makeWrapper
+, fontconfig
+, noto-fonts
+, libclang
+}:
+
+rustPlatform.buildRustPackage rec {
+  pname = "neothesia" + (if cli then "-cli" else "");
```
Comment (lilyinstarlight) [inline pkgs/applications/audio/neothesia/default.nix:21]:
```suggestion
  pname = "neothesia" + (lib.optionalString cli "-cli");
```

We have a helper function for this

File: pkgs/applications/audio/neothesia/default.nix [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  version = "unstable-2023-09-23";
+
+  src = fetchFromGitHub {
+    owner = "PolyMeilex";
+    repo = "Neothesia";
+    rev = "d724865059e876f14a3b92ba18a1c5dbe3485131";
+    hash = "sha256-CnVfv0iCuUgZcmtpxg8+wgG7r8HsLdKgjCbbNu5o76c=";
+  };
+
+  strictDeps = true;
```
Comment (lilyinstarlight) [inline pkgs/applications/audio/neothesia/default.nix:31]:
```suggestion
```

This is implicit for `buildRustPackage`

File: pkgs/applications/audio/neothesia/default.nix [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  cargoLock = {
+    lockFile = "${src}/Cargo.lock";
+    outputHashes = {
+      "mpeg_encoder-0.2.1" = "sha256-+BNZZ1FIr1374n8Zs1mww2w3eWHOH6ENOTYXz9RT2Ck=";
+    };
+  };
+
+  cargoBuildFlags = [ "--package" pname ];
+
+  #env.LIBCLANG_PATH = "${lib.makeLibraryPath [ libclang.lib ]}";
```
Comment (lilyinstarlight) [inline pkgs/applications/audio/neothesia/default.nix:62]:
```suggestion
```

File: pkgs/applications/audio/neothesia/default.nix [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+  postInstall = (if cli then "" else ''
+    install -D flatpak/com.github.polymeilex.neothesia.desktop -t $out/share/applications
+    install -D flatpak/com.github.polymeilex.neothesia.png -t $out/share/icons/hicolor/256x256/apps
+  '');
```
Comment (lilyinstarlight) [inline pkgs/applications/audio/neothesia/default.nix:67]:
```suggestion
  postInstall = lib.optionalString (!cli) ''
    install -D flatpak/com.github.polymeilex.neothesia.desktop -t $out/share/applications
    install -D flatpak/com.github.polymeilex.neothesia.png -t $out/share/icons/hicolor/256x256/apps
  '';
```

File: pkgs/applications/audio/neothesia/default.nix [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+      When executing the program, you input the path to the MIDI file as the first argument. It will then render the notes flying towards a piano where a key lights up when a note is played. This is encoded into a video file.
+    '' else ''
+      Neothesia is a GUI program which can play MIDI files, visualising the notes on a piano it plays for you or which you play with MIDI input. Before the notes are played, they fly towars the piano.
+
+      It plays the notes through the sound output too. The default instrument sound can be changed by loading SoundFont files.
+
+      Neothesia may help you to learn how to play on the piano.
+    '');
+    homepage = "https://github.com/PolyMeilex/Neothesia";
+    license = licenses.gpl3;
```
Comment (lilyinstarlight) [inline pkgs/applications/audio/neothesia/default.nix:89]:
```suggestion
    license = licenses.gpl3Only;
```

We prefer either `gpl3Plus` if the project is explicit later versions of GPL can apply or `gpl3Only` when they do not. This case seems to be `gpl3Only`

File: pkgs/applications/audio/neothesia/default.nix [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    owner = "PolyMeilex";
+    repo = "Neothesia";
+    rev = "d724865059e876f14a3b92ba18a1c5dbe3485131";
+    hash = "sha256-CnVfv0iCuUgZcmtpxg8+wgG7r8HsLdKgjCbbNu5o76c=";
+  };
+
+  strictDeps = true;
+  nativeBuildInputs = [
+    pkg-config
+    cmake
```
Comment (lilyinstarlight) [inline pkgs/applications/audio/neothesia/default.nix:34]: What is cmake actually doing here? I can't tell that it gets used

File: pkgs/applications/audio/neothesia/default.nix [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+, alsa-lib
+, atkmm
+, gtk3
+, wrapGAppsHook
+, ffmpeg-full
+, vulkan-loader
+, makeWrapper
+, fontconfig
+, noto-fonts
+, libclang
```
Comment (lilyinstarlight) [inline pkgs/applications/audio/neothesia/default.nix:17]:
```suggestion
```

Unused args

**Review by lilyinstarlight — COMMENTED** [review]
This looks pretty good! The `out/video.mp4` file needs to be removed and I left a few more comments

Also regarding https://github.com/PolyMeilex/Neothesia/issues/77#issuecomment-1732568007, it looks like cosmic-text has since made a release

Thank you for this!

**Comment by auroraanna** [comment]
Hmm, on my latest build, Neothesia crashes.
```console
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', /build/cargo-vendor-dir/wgpu-hal-0.17.1/src/gles/egl.rs:813:77
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: queue 0x55df0e934650 destroyed while proxies still attached:
  xdg_wm_base@21 still attached
  wl_output@15 still attached
  wl_output@14 still attached
  wl_seat@13 still attached
  xdg_activation_v1@12 still attached
  wp_viewporter@11 still attached
  zwp_text_input_manager_v3@10 still attached
  zwp_pointer_constraints_v1@9 still attached
  zwp_relative_pointer_manager_v1@8 still attached
  zxdg_decoration_manager_v1@7 still attached
  wl_subcompositor@6 still attached
  wl_compositor@5 still attached
  wl_shm@4 still attached
  wl_registry@2 still attached
```

File: pkgs/applications/audio/neothesia/default.nix (hunk 6 above)
Comment (auroraanna) [reply pkgs/applications/audio/neothesia/default.nix:31]: hmm, I don't remember anymore why I put it there but I didn't get any errors when building without it so I guess it's not needed.

Comment (auroraanna) [reply pkgs/applications/audio/neothesia/default.nix]: not getting any errors without atkmm anymore either. pretty sure it was needed at some point.

**Comment by lilyinstarlight** [comment]
> Hmm, on my latest build, Neothesia crashes.

Hmmm, I'll try to investigate at some point myself too, because that does seem weird

**Comment by auroraanna** [comment]
So, neothesia has the new cosmic-text version now but it still crashes for me. Even tried it on Xorg and found that libxkbcommon was missing there so I added that.

**Comment by auroraanna** [comment]
oops, i forgot someone already added neothesia under by-name and thought i did that

**Pull request closed without merging**